### PR TITLE
DAO example in Motoko

### PR DESCRIPTION
**Pull request by chenyan-dfinity**

### CODE REVIEW

**Comment by crusso** [comment]
Will this be merged with #142. It'll need a README (maybe just symlink the other one if that works)

Do the rust/motoko examples implement exactly the same interface?

Can they use your ic-repl test to the rust example?

File: motoko/dao/src/Main.mo [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            switch (await execute_proposal(proposal)) {
+            case (#ok) { update_proposal_state(proposal, #succeeded); };
+            case (#err(err)) { update_proposal_state(proposal, #failed(err)); };
+            };
+        };
+    };
+
+    /// Execute the given proposal
+    func execute_proposal(proposal: Types.Proposal) : async Types.Result<(), Text> {
+        // unimplemented until raw call is supported
```
Comment (crusso) [inline motoko/dao/src/Main.mo:200]: Ha!

File: motoko/dao/tests/proposal.test.sh [hunk 2]
```diff
@@ -0,0 +1,148 @@
+#!ic-repl
+load "prelude.sh";
```
Comment (crusso) [inline motoko/dao/tests/proposal.test.sh:2]: ic-repl looks nice - I should actually use it...

Comment (crusso) [reply motoko/dao/tests/proposal.test.sh:2]: But it also looks very sequential - are all calls made synchronous?

File: motoko/dao/src/Types.mo [hunk 3]
```diff
@@ -0,0 +1,65 @@
+import Result "mo:base/Result";
```
Comment (crusso) [inline motoko/dao/src/Types.mo:1]: Does this mean Rust and Motoko will have different candid interface with #Err vs #err?

File: motoko/dao/tests/proposal.test.sh (hunk 2 above)
Comment (chenyan-dfinity) [reply motoko/dao/tests/proposal.test.sh:2]: All calls have an await, so the script is sequential. Do you have a use case to make concurrent calls?

File: motoko/dao/src/Types.mo (hunk 3 above)
Comment (chenyan-dfinity) [reply motoko/dao/src/Types.mo:1]: Yes

File: motoko/dao/src/Types.mo [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import List "mo:base/List";
+import Principal "mo:base/Principal";
+
+module {
+  public type Result<T, E> = Result.Result<T, E>;
+  public type Account = { owner : Principal; tokens : Tokens };
+  public type Proposal = {
+    id : Nat;
+    votes_no : Tokens;
+    voters : List.List<Principal>;
```
Comment (chenyan-dfinity) [inline motoko/dao/src/Types.mo:13]: Besides `Err` vs `err`, this is another difference in the did file. Rust is using `vec principal`, and Motoko uses `opt record { principal; list }`.

The Rust implementation can use the same ic-repl scripts, except for the above differences.

**Review by ggreif — APPROVED** [review]
Go for it!

**Pull request merged**